pkg/storage/chunk/client/ibmcloud/trusted_profile_authentication_provider.go

```diff
@@ -0,0 +1,159 @@
+)
+
+// TrustedProfileProvider Struct
+// This implements Provider interface from https://github.com/IBM/ibm-cos-sdk-go
+type TrustedProfileProvider struct {
+	// Name of Provider
+	providerName string
+
+	// Type of Provider - SharedCred, SharedConfig, etc.
+	providerType string
+
+	// Authenticator implements an IAM-based authentication schema
+	authenticator *core.ContainerAuthenticator
+
+	// Error
+	ErrorStatus error
+}
+
+// NewTrustedProfileProvider allows the creation of a custom IBM IAM Provider
+// Parameters:
+//
+//	Provider Name
+//	AWS Config
+//	Trusted Profile Name
+//	Trusted Profile ID
+//	Compute Resource Token File Path
+//	IBM IAM Authentication Server Endpoint
+//
+// Returns:
+//
+//	TrustedProfileProvider
+func NewTrustedProfileProvider(providerName string, trustedProfileName,
+	trustedProfileID, crTokenFilePath, authEndpoint string) *TrustedProfileProvider {
+	provider := new(TrustedProfileProvider)
+
+	provider.providerName = providerName
+	provider.providerType = "oauth"
+
+	if trustedProfileName == "" && trustedProfileID == "" {
+		provider.ErrorStatus = awserr.New("trustedProfileNotFound", "either Trusted profile name or id must be provided", nil)
+		level.Debug(log.Logger).Log("msg", provider.ErrorStatus)
+
+		return provider
+	}
+
+	if crTokenFilePath == "" {
+		provider.ErrorStatus = awserr.New("crTokenFilePathEmpty", "must supply cr token file path", nil)
+		level.Debug(log.Logger).Log("msg", provider.ErrorStatus)
+
+		return provider
+	} else {
+		if _, err := os.Stat(crTokenFilePath); errors.Is(err, os.ErrNotExist) {
+			provider.ErrorStatus = awserr.New("crTokenFileNotFound", "no such file", err)
+			level.Debug(log.Logger).Log("msg", "no such file", "err", err)
+
+			return provider
+		}
+	}
+
+	if authEndpoint == "" {
+		authEndpoint = defaultCOSAuthEndpoint
+		level.Debug(log.Logger).Log("msg", "using default auth endpoint", "endpoint", authEndpoint)
+	}
+
+	authenticator, err := core.NewContainerAuthenticatorBuilder().
+		SetIAMProfileName(trustedProfileName).
+		SetIAMProfileID(trustedProfileID).
+		SetCRTokenFilename(crTokenFilePath).
+		SetURL(authEndpoint).
+		Build()
+
+	if err != nil {
+		provider.ErrorStatus = awserr.New("errCreatingAuthenticatorClient", "failed to setup new Trusted Profile Authenticator client", err)
+		level.Debug(log.Logger).Log("msg", provider.ErrorStatus)
+
+		return provider
+	}
+
+	provider.authenticator = authenticator
+
+	return provider
+}
+
+// IsValid ...
+// Returns:
+//
+//	TrustedProfileProvider validation - boolean
+func (p *TrustedProfileProvider) IsValid() bool {
+	return nil == p.ErrorStatus
+}
+
+// Retrieve ...
+// Returns:
+//
+//	Credential values
+//	Error
+func (p *TrustedProfileProvider) Retrieve() (credentials.Value, error) {
+	if p.ErrorStatus != nil {
+		level.Debug(log.Logger).Log("msg", p.ErrorStatus)
+
+		return credentials.Value{ProviderName: p.providerName}, p.ErrorStatus
+	}
+
+	tokenValue, err := p.authenticator.GetToken()
```

Comment:
add a comment to explain that this gets a token or generate a new one if expired or there is none that are already retrieved.

Reply:
+1